### ISSUE TITLE
[Flutter drive] cannot find file when running on web

### ISSUE THREAD

**Issue by cedvdb** [flutter/flutter#102469]
Repro:

1. clone https://github.com/cedvdb/flutter_repros/tree/integration_test_failing_imports (integration_test_failing_imports branch)
2. run pub get
3. Run test in web `flutter drive --driver=test/integration_test_driver.dart --target=test/acceptance/tests/feature/feature_acceptance_test.dart --no-pub --device-id=web-server --no-headless`

(optionally run test on android)
Run test in android `flutter drive --driver=test/integration_test_driver.dart --target=test/acceptance/tests/feature/feature_acceptance_test.dart --no-pub`

# Result

```
org-dartlang-app:/feature_acceptance_test.dart:6:8: Error: Error when reading 'org-dartlang-app:/utils/add.dart': File not found
import '../../utils/add.dart';
```

#Expected  Result

For it to run like on android

**Comment by darshankawar** [comment]
@cedvdb 
Not sure but I am wondering if it has to do with the hierarchy of your project and specially the `test` folder structure and if web handles imports like this in a different way `import '../../../../test/acceptance/utils/add.dart';
`
It looks similar case as `dart:io` imports are not supported in flutter web and could be related to it. Also, since `int add(int a, int b) => a + b;` seems to be coming from dart and hence the web part may not be handling it as simiilar to https://github.com/flutter/flutter/issues/39998. Please check and confirm.

**Comment by cedvdb** [comment]
@darshankawar What exactly do you need me to check and confirm ? It's not clear.

As for the folder structure, it was my assumption as well.

**Comment by darshankawar** [comment]
> Ence why I see the error `Error when reading 'org-dartlang-app:/...` That'd certainly make sense.

That's my assumption as well and hence I think that it could be similar to the issue I linked earlier. https://github.com/flutter/flutter/issues/39998

For now, since I do see the same behavior / error after running your code, I'll go ahead and label it for further insights from the team on the intended behavior and whether the root-cause is same and duplicate of    https://github.com/flutter/flutter/issues/39998


<details>
<summary> stable, master flutter doctor -v </summary>

```

[✓] Flutter (Channel stable, 2.10.5, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 2.10.5 at /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5464c5bac7 (12 hours ago), 2022-04-18 09:55:37 -0700
    • Engine revision 57d3bac3dd
    • Dart version 2.16.2
    • DevTools version 2.9.2

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[✓] Flutter (Channel master, 2.13.0-0.0.pre.732, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 2.13.0-0.0.pre.732 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fa96ddd427 (59 minutes ago), 2022-04-25 21:39:07 -0700
    • Engine revision 4df9cf2705
    • Dart version 2.18.0 (build 2.18.0-56.0.dev)
    • DevTools version 2.12.2
    
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel beta, 2.13.0-0.1.pre, on macOS 12.2.1 21D62 darwin-x64,
    locale en-GB)
[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
[✓] Chrome - develop for the web
[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
[✓] VS Code (version 1.62.0)
[✓] Connected device (3 available)
[✓] HTTP Host Availability



```

</details>

**Comment by cedvdb** [comment]
I'd hope this won't be labeled as a duplicate as it is blocking acceptance tests and the other issue seems to be a bigger one.

Maybe switching putting our test folder into the `package` folder with its own pubspec.yaml could work as a workaround.

**Comment by yjbanov** [comment]
This looks like a bug either in the tool or in the compiler.

**Comment by yjbanov** [comment]
It may be because the entrypoint for the target is under the `test/` directory. As a workaround, try using the file structure described in https://docs.flutter.dev/cookbook/testing/integration/introduction

**Comment by cedvdb** [comment]
@yjbanov I confirm that moving /test to /lib/test is a valid workaround. Thanks

**Comment by cedvdb** [comment]
@yjbanov the workaround you proposed does not work. The issue is with the `../..` in the import paths, simply renaming the top folder has no bearing in that error. It seems like you misunderstood the issue, please reconsider the label P5 in light of this.

The only (annoying) workaround that was found was to put the test files under the `lib` folder, but then the dev dependencies have to be moved to normal dependencies to have imports work well with intellisens (I guess dev deps are not supposed to be imported there so they do not appear).

**Comment by cedvdb** [comment]
@darshankawar Could you add the appropriate labels as it is found in flutter 3 ? I think there is a label regarding dart2js missing too ? And the low priority here is beyond comprehension to me but so be it.

**Comment by maheshmnj** [comment]
@cedvdb I ran the tests on android and it seems to succeed with relative imports, But when running on the web the test fails, I  moved `feature_acceptance_test.dart` and `add.dart` under test and the test succeeds. This issue is reproducible on both the latest stable and the master channel.

I do see the same has been mentioned above by @darshankawar

> Not sure but I am wondering if it has to do with the hierarchy of your project and specially the test folder structure and if web handles imports like this in a different way import '../../../../test/acceptance/utils/add.dart';
It looks similar case as dart:io

### Android test passes with relative imports

```
mahesh@Maheshs-MacBook-Air-M1 flutter_repros % flutter drive --driver=test/integration_test_driver.dart --target=test/acceptance/tests/feature/feature_acceptance_test.dart --no-pub
Warning: Mapping new ns http://schemas.android.com/repository/android/common/02 to old ns http://schemas.android.com/repository/android/common/01
Warning: Mapping new ns http://schemas.android.com/repository/android/generic/02 to old ns http://schemas.android.com/repository/android/generic/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/addon2/02 to old ns http://schemas.android.com/sdk/android/repo/addon2/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/repository2/02 to old ns http://schemas.android.com/sdk/android/repo/repository2/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/sys-img2/02 to old ns http://schemas.android.com/sdk/android/repo/sys-img2/01
Warning: Mapping new ns http://schemas.android.com/repository/android/common/02 to old ns http://schemas.android.com/repository/android/common/01
Warning: Mapping new ns http://schemas.android.com/repository/android/generic/02 to old ns http://schemas.android.com/repository/android/generic/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/addon2/02 to old ns http://schemas.android.com/sdk/android/repo/addon2/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/repository2/02 to old ns http://schemas.android.com/sdk/android/repo/repository2/01
Warning: Mapping new ns http://schemas.android.com/sdk/android/repo/sys-img2/02 to old ns http://schemas.android.com/sdk/android/repo/sys-img2/01
Running Gradle task 'assembleDebug'...                             37.1s
✓  Built build/app/outputs/flutter-apk/app-debug.apk.
Installing build/app/outputs/flutter-apk/app.apk...                883ms
W/.flutter_repro( 1513): Accessing hidden method
Landroid/app/ActivityThread;->currentActivityThread()Landroid/app/ActivityThread; (greylist, linking, allowed)
W/.flutter_repro( 1513): Accessing hidden method
Landroid/app/ActivityThread;->getActivity(Landroid/os/IBinder;)Landroid/app/Activity; (greylist, linking, allowed)
VMServiceFlutterDriver: Connecting to Flutter application at http://127.0.0.1:61462/Kx-qdyjZBh8=/
VMServiceFlutterDriver: Isolate found with number: 3222012541120307
VMServiceFlutterDriver: Isolate is paused at start.
VMServiceFlutterDriver: Attempting to resume isolate
I/flutter ( 1513): 00:00 +0: end-to-end test tap on the floating action button, verify counter
VMServiceFlutterDriver: Connected to Flutter application.
I/flutter ( 1513): 00:01 +1: (tearDownAll)
All tests passed.
I/flutter ( 1513): 00:01 +2: All tests passed!
```

### Tests on the web fails with relative imports

```
mahesh@Maheshs-MacBook-Air-M1 flutter_repros % flutter drive --driver=test/integration_test_driver.dart --target=test/acceptance/tests/feature/feature_acceptance_test.dart -d web-server 
Running "flutter pub get" in flutter_repros...                     343ms
Launching test/acceptance/tests/feature/feature_acceptance_test.dart on Web Server in debug mode...
org-dartlang-app:/feature_acceptance_test.dart:6:8: Error: Error when reading 'org-dartlang-app:/utils/add.dart':
File not found
import '../../utils/add.dart';
       ^
org-dartlang-app:/feature_acceptance_test.dart:19:14: Error: Method not found: 'add'.
      expect(add(1, 2), equals(3));
             ^^^
Waiting for connection from debug service on Web Server...         10.6s
Failed to compile application.
Application exited before the test started. Check web driver logs for possible application-side errors.

mahesh@Maheshs-MacBook-Air-M1 flutter_repros % flutter drive --driver=test/integration_test_driver.dart --target=test/feature_acceptance_test.dart -d web-server
Running "flutter pub get" in flutter_repros...                     386ms
Launching test/feature_acceptance_test.dart on Web Server in debug mode...
Waiting for connection from debug service on Web Server...         10.5s
test/feature_acceptance_test.dart is being served at http://localhost:63600
The web-server device requires the Dart Debug Chrome extension for debugging. Consider using the Chrome or Edge
devices for an improved development workflow.
Unable to start a WebDriver session for web testing.
Make sure you have the correct WebDriver server (e.g. chromedriver) running at 4444.
For instructions on how to obtain and run a WebDriver server, see:
https://flutter.dev/docs/testing/integration-tests#running-in-a-browser


mahesh@Maheshs-MacBook-Air-M1 flutter_repros % flutter drive --driver=test/integration_test_driver.dart --target=test/feature_acceptance_test.dart -d web-server
Running "flutter pub get" in flutter_repros...                     270ms
Launching test/feature_acceptance_test.dart on Web Server in debug mode...
Waiting for connection from debug service on Web Server...         10.6s
test/feature_acceptance_test.dart is being served at http://localhost:63710
The web-server device requires the Dart Debug Chrome extension for debugging. Consider using the
Chrome or Edge devices for an improved development workflow.
All tests passed.
Application finished.
```

<details>
<summary>flutter doctor -v (mac)</summary>

```
[✓] Flutter (Channel stable, 3.0.2, on macOS 12.4 21F79 darwin-arm, locale en-IN)
    • Flutter version 3.0.2 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision cd41fdd495 (5 days ago), 2022-06-08 09:52:13 -0700
    • Engine revision f15f824b57
    • Dart version 2.17.3
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-32, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.42.0

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 102.0.5005.61

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```
[✓] Flutter (Channel master, 3.1.0-0.0.pre.1237, on macOS 12.4 21F79 darwin-arm, locale en-IN)
    • Flutter version 3.1.0-0.0.pre.1237 at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6062e40510 (64 minutes ago), 2022-06-12 23:53:08 -0400
    • Engine revision f80294b57e
    • Dart version 2.18.0 (build 2.18.0-180.0.dev)
    • DevTools version 2.14.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-32, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.42.0

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 102.0.5005.61

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

**Comment by ominibyte** [comment]
Any news on this? or workaround? This is a huge blocker.

**Comment by Maarten0110** [comment]
Still no news? I too don't get why this has the lowest priority. This issue is blocking us from running our integration test suite for web because we have a bunch of integration test related utility functions that we need to import in each test..

**Comment by maheshmnj** [comment]
Everyone who is affected by this issue, Please upvote the [original issue report](https://github.com/flutter/flutter/issues/102469#issue-1213690308). The number of upvotes helps to understand the no of users affected by this issue. The number of upvotes is one among many factors to bump up the priority.

**Comment by tee-theeraporn-hugeman** [comment]
I has this issue. I fix by running in profile mode or release mode because in my. understanding if you import file with `../` in dev mode that' file is not include to the test if you running in profile or release mode after running test it's compile you code include another folder. that's why can fix my problem

before i run this command
`flutter drive --driver=test_driver/integration_test.dart --target=integration_test/b/foo_test.dart -d chrome`

after
`flutter drive --driver=test_driver/integration_test.dart --target=integration_test/b/foo_test.dart -d chrome --release`

my code
```
my folder structure
integration_test/
├── a/
          ├── foo_test.dart
├── b/
          ├── bar_test.dart
          
my code in bar_test.dart need to import file from folder a/

import '../foo_test.dart';  <--- has issue if import another folder
```

_P.S. so sorry in my bad english but i hope to solve your problem_

> Still no news? I too don't get why this has the lowest priority. This issue is blocking us from running our integration test suite for web because we have a bunch of integration test related utility functions that we need to import in each test..
> 
> Edit: Relative import without dots `../` seems so work, so for now we can proceed by having our util file in the same folder as the web test files (not ideal and probably not a solution for everyone):
> 
> file structure:
> 
> ```
> integration_test/
>     mobile/
>         foo_test.dart
>         bar_test.dart
>     web/
>         foo_test.dart
>         bar_test.dart
>         util.dart
> ```
> 
> import in the `web/` test files:
> 
> ```
> import 'util.dart';
> ```

**Comment by maxeema** [comment]
The issue is still reproduced in Flutter 3.7.1

**Comment by AsturaPhoenix** [comment]
So, adding the project root path to the compiler seems to make it able to resolve these relative imports, but I'm not entirely clear exactly why it works or where to add tests. If CI reveals something obvious, I can give it a further shot. Otherwise, PR #120944 is a one-liner that seems to fix it for me so far (only for my project, without any rigorous testing).

**Comment by AsturaPhoenix** [comment]
*tl;dr:* The next thing to investigate is probably that something is stripping away the relative path elements from the import directives.

So I've run a different test with my hack, which leads me to *not* recommend that we move forward with my hack, but may shed some further light on a root cause. Essentially my hack takes advantage of noticing that the error messages complained about not being able to find the imports under the root directory. (e.g. for the OP's case of `../../util/add.dart`, `Error when reading 'org-dartlang-app:/utils/add.dart': File not found`.) Thus, adding the actual project root to the project roots passed to the compiler kinda works, since these relative paths are likely to `..` up to the project root anyway.

However, a) it looks like the compiler is simply stripping away the relative path elements, which is probably a bug and the root cause, and b) this hack doesn't always work, if the relative directory doesn't path up all the way to the project root (or the immediate parent or `test` folder, which are also added as project roots to the compiler).

Example, if I have directory structure
```
integration_test/
    foo_test.dart
test/
    src/
        foo_dep.dart
        bar_dep.dart
        baz_dep.dart
```

and `integration_test/foo_test.dart`
```dart
import '../test/src/foo_dep.dart';
import './test/src/bar_dep.dart';
import '../../test/src/baz_dep.dart';
```

with my patch, the test happily succeeds since it's able to find all those files under the project root.

If on the other hand my directory structure is
```
integration_test/
    foo/
        bar_test.dart
        baz.dart
```

and I have `integration_test/foo/foo_test.dart`
```dart
import '../foo/baz.dart';
```

even though the IDE and Android build think everything should be fine, `flutter drive -t integration_test/foo/foo_test.dart -d chrome` fails with the familiar
```
org-dartlang-app:/bar_test.dart:1:8: Error: Error when reading 'org-dartlang-app:/foo/baz.dart': File not found
import '../foo/baz.dart';
       ^
```

**Comment by jithware** [comment]
When running on [web](https://docs.flutter.dev/cookbook/testing/integration/introduction#5b-web), I am seeing the same error:
```
flutter drive --no-pub --driver=test_driver/integration_test.dart --target=integration_test/app_test.dart -d chrome 
Launching integration_test/app_test.dart on Chrome in debug mode...
org-dartlang-app:/app_test.dart:11:8: Error: Error when reading 'org-dartlang-app:/test/home_widget_test.dart':
File not found
import '../test/home_widget_test.dart';
```
Works fine when running on [mobile](https://docs.flutter.dev/cookbook/testing/integration/introduction#5a-mobile):
```
flutter test integration_test/app_test.dart 
...
02:12 +1: All tests passed!  
```

**Comment by Mohammad3id** [comment]
Sill no fix for this issue?
